### PR TITLE
Bug: fix to not invoke `transfer` for call related opcodes (except `CALL`)

### PR DESCRIPTION
**Pull request by silathdiir**
Close https://github.com/privacy-scaling-explorations/zkevm-specs/issues/334

Related circuit PR https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/973

As issue described, there should be no `transfer` invocation for call related opcodes (except [CALL](https://github.com/ethereum/go-ethereum/blob/master/core/vm/evm.go#L167)) - [CALLCODE](https://github.com/ethereum/go-ethereum/blob/master/core/vm/evm.go#L253), [DELEGATECALL](https://github.com/ethereum/go-ethereum/blob/master/core/vm/evm.go#L301) and [STATICCALL](https://github.com/ethereum/go-ethereum/blob/master/core/vm/evm.go#L340).

~~And callee balance is also needed for checking account existence below (for now). Suppose if it could be fixed to replace with [non-existing proofs](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/907) after using in the [BALANCE circuit](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/683).~~

### CODE REVIEW

File: src/zkevm_specs/evm/execution/callop.py [hunk 1]
```diff
@@ -108,21 +108,27 @@ def callop(instruction: Instruction):
...
-        caller_address, callee_address, value, callee_reversion_info
-    )
+    if is_call == 1:
+        # Verify transfer only for CALL opcode.
+        _, (_, callee_balance) = instruction.transfer(
+            caller_address, callee_address, value, callee_reversion_info
+        )
+    else:
+        # Get callee balance for CALLCODE, DELEGATECALL and STATICCALL opcodes.
+        callee_balance = instruction.account_read(callee_address, AccountFieldTag.Balance)
```
Comment (ed255) [inline src/zkevm_specs/evm/execution/callop.py:118]: According to the CALLCODE implementation you linked in https://github.com/privacy-scaling-explorations/zkevm-specs/issues/334#issue-1481042446
For CALLCODE you should constraint that `value <= callee_balance` right?
I think I didn't see this constraint.

**Comment by lispc** [comment]
@CPerezz don't need to be so strict... 

Sometimes we prefer perfect PR, sometimes we prefer urgent PR. For the latter, as long as a PR is better than before, it can be accepted. 

This issue is a bit urgent both for pse/zkevm-chain and scroll circuits, related to https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/975 and https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/969

**Comment by CPerezz** [comment]
Hey @lispc .

I'm not sure if I'm interpreting your message correctly. But let me be clear. I think that for a codebase this big and complex, being "strict" if you want to call it like this, is not crazy and rather what we all should be when we review.

I don't think I have asked anything crazy considering that we just need to remove 3 lines of code to include the existance lookup swapping it by the balance, codehash and value checks..

In any case, @ed255 I think is in the same line as I am. We cannot merge anything just for the purpose of being fast. And, instead, we should try to fix things propperly (without of course affecting the overall progress of the project). 

Summarizing. Using here the non-existance lookup is a task we need to do in the close future and takes 1 day of work MAXIMUM. So why leaving more tech-debt behind when we can already fix it quickly here and avoid more TODO's??

**Comment by lispc** [comment]
After some discussion we choose to let non-existing proofs be a TODO (future PR). @silathdiir 

Anyway 1 day dev time does not mean 1 day diff in wall clock delivery time. Developers have other tasks to do, reviewers ditto.

**Review by lispc — APPROVED** [review]

**Comment by aguzmant103** [comment]
Hi @silathdiir ! Just checking in, are you good to apply the changes to merge this PR?

**Comment by silathdiir** [comment]
> Hi @silathdiir ! Just checking in, are you good to apply the changes to merge this PR?

Yes. Will update this PR according to code review in this weekend. Thanks.

**Comment by aguzmant103** [comment]
Thank you!!

File: src/zkevm_specs/evm/execution/callop.py [hunk 2]
```diff
@@ -108,21 +108,27 @@ def callop(instruction: Instruction):
...
     # Verify gas cost
     callee_nonce = instruction.account_read(callee_address, AccountFieldTag.Nonce)
     callee_code_hash = instruction.account_read(code_address, AccountFieldTag.CodeHash)
     is_empty_code_hash = instruction.is_equal(
         callee_code_hash, instruction.rlc_encode(EMPTY_CODE_HASH, 32)
     )
+    # TODO:
+    # Suppose to fix to use non-existing proofs for account existence after it
+    # has been used for one opcode in zkevm-circuit.
+    # https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/907
```
[the remark this answers is not in the document]
Comment (silathdiir) [reply src/zkevm_specs/evm/execution/callop.py]: Deleted the TODO comment. And fixed to pass `callee_exists` via auxiliary witness data.

File: src/zkevm_specs/evm/execution/callop.py [hunk 3]
```diff
@@ -108,21 +108,27 @@ def callop(instruction: Instruction):
...
     is_account_empty = (
-        instruction.is_zero(callee_nonce)
-        * instruction.is_zero(callee_balance_prev)
-        * is_empty_code_hash
+        instruction.is_zero(callee_nonce) * instruction.is_zero(callee_balance) * is_empty_code_hash
```
[the remark this answers is not in the document]
Comment (silathdiir) [reply src/zkevm_specs/evm/execution/callop.py]: Has been replaced with `callee_exists`.

File: src/zkevm_specs/evm/execution/callop.py (hunk 1 above)
Comment (silathdiir) [reply src/zkevm_specs/evm/execution/callop.py]: Fixed to add a constraint to verify caller balance is greater than or equal to stack `value` for both `CALL` and `CALLCODE` opcodes as below:
```
if is_call + is_callcode == 1:
    value_lt_caller_balance, value_eq_caller_balance = instruction.compare_word(
        value, caller_balance
)
instruction.constrain_zero(1 - value_lt_caller_balance - value_eq_caller_balance)
```

**Comment by silathdiir** [comment]
Hi @CPerezz and @ed255 , I updated this PR according to above comments. Could you help review again when you have time? Thanks.

**Review by CPerezz — APPROVED** [review]
LGTM! 

Thanks for this work @silathdiir :partying_face:

**Pull request merged**